File: Copy List with Random Pointer.java
A linked list of length n is given such that each node contains an additional random pointer, which could point to any node in the list, or null.

Construct a deep copy of the list. The deep copy should consist of exactly n brand new nodes, 
where each new node has its value set to the value of its corresponding original node. 

https://leetcode.com/explore/learn/card/linked-list/213/conclusion/1229/


class Solution 
{
    public Node copyRandomList(Node head) 
    {
        if(head == null) return null;
        
        Node ptr = head;
        
        // Make a copy of each node in the original list and put them in a map.
        // Using a map to instantly look up old node -> new node.
        Map<Node, Node> nodeMap = new HashMap<>();
        while(ptr != null)
        {
            nodeMap.put(ptr, new Node(ptr.val));
            ptr = ptr.next;
        }
        
        // Reset ptr back to the head of the original list.
        // Go through the map again to connect next and random.
        ptr = head;
        while(ptr != null)
        {
            Node copyPtr = nodeMap.get(ptr);
            copyPtr.next = nodeMap.get(ptr.next);
            copyPtr.random = nodeMap.get(ptr.random);
            ptr = ptr.next;
        }
        return nodeMap.get(head);
    }
}

File: Copy List with Random Pointer.py
https://leetcode.com/problems/copy-list-with-random-pointer/

"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head:
            return head

        old_to_new_mapping = {}
        ptr = head
        new_head = dummy = None
        while ptr:
            val = ptr.val

            if not new_head:
                new_head = Node(x=val)
                dummy = new_head
            else:
                dummy.next = Node(x=val)
                dummy = dummy.next

            old_to_new_mapping[ptr] = dummy
            ptr = ptr.next

        ptr = head
        while ptr:
            if ptr.random:
                old_to_new_mapping[ptr].random = old_to_new_mapping[ptr.random]

            ptr = ptr.next

        return new_head
